[API Implementation]: Expose general purpose Crc32 APIs

deeprobin opened the pull request:
Proposal implementation of #2036 (closes #2036)

## Proposed API
```cs
namespace System.Numerics
{
    public static class BitOperations
    {
        public static uint Crc32C(uint crc, byte data);
        public static uint Crc32C(uint crc, ushort data);
        public static uint Crc32C(uint crc, uint data);
        public static uint Crc32C(uint crc, ulong data);
    }
}
```

## Current state of implementation
* [x] Ref Assembly
* [x] Documentation
* [x] Implementation
  * [x] Hardware Intrinsics
     * [x] SSE 4.2
     * [x] SSE 4.2 x64 
     * [x] ARM
     * [x] ARM x64 
  * [x] Software Fallback 
* [x] Tests

## Usage of Hardware Intrinsics

```h
uint32_t __crc32b (uint32_t a, uint8_t b); // SSE 4.2
uint32_t __crc32cb (uint32_t a, uint8_t b); // ARM

unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v); // SSE 4.2
uint32_t __crc32ch (uint32_t a, uint16_t b); // ARM

unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v); // SSE 4.2
uint32_t __crc32ch (uint32_t a, uint32_t b); // ARM

unsigned __int64 _mm_crc32_u64 (unsigned __int64 crc, unsigned __int64 v); // SSE 4.2 x64
uint32_t __crc32cd (uint32_t a, uint64_t b); // ARM x64
```

## Software Fallback 
```cs
            foreach (byte b in bytes)
            {
                int tableIndex = (int)((crc ^ b) & 0xFF);
                crc = s_crcTable[tableIndex] ^ (crc >> 8);
            }
```

/cc @tannergooding

deeprobin committed:
Expose Crc32 Implementation and implement Crc32 software fallback

deeprobin committed:
Fix signature of BitOperation Crc32 methods

deeprobin committed:
Implement common test cases

deeprobin committed:
Add X64 Intrinsic support for Crc32(uint crc, ulong data)

deeprobin committed:
Add documentation comments

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +708,167 @@ public static nuint RotateRight(nuint value, int offset)
+            // Software fallback
+
+            // Implicit byte[] to ReadOnlySpan<byte> conversion
+            ReadOnlySpan<byte> bytes = BitConverter.GetBytes(data);
```
@tannergooding Is this a good approach or should I use `Unsafe.As`?

deeprobin commented:
> 
> Note regarding the `new-api-needs-documentation` label:
> 

Tag can be removed, documentation is implemented :smile:

deeprobin committed:
Remove Hwintrinsic doc-comment

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +708,166 @@ public static nuint RotateRight(nuint value, int offset)
+            if (Sse42.X64.IsSupported)
+            {
+                // unsigned __int64 _mm_crc32_u64 (unsigned __int64 crc, unsigned __int64 v)
+                return (uint)Sse42.X64.Crc32(crc, data);
```
Some tests fail here due casting ulong to uint. Is there a Sse42 X64 equivalent for calculating a Crc checksum to uint?

deeprobin replied:
There is no Crc32C support which returns `uint32_t` with data paremter of type `__int64`.
See [Intel® Intrinsics Guide](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#techs=MMX,SSE,SSE2,SSE3,SSSE3,SSE4_1,SSE4_2,AVX,AVX2,FMA,AVX_VNNI,AVX_512,KNC,AMX,SVML,Other&text=crc)

deeprobin replied:
I removed this.

saucecontrol replied:
`_mm_crc32_u64` returns `__int64`, of which only the low 32 bits are populated.  It can be truncated safely.

deeprobin replied:
@saucecontrol Does this return the Castagnoli variant?

saucecontrol replied:
Yes, it's the same instruction as emitted by `_mm_crc32_u32`, just with different encoding to indicate it operates on 64 bits instead of 32.  See https://www.felixcloutier.com/x86/crc32.

deeprobin replied:
@saucecontrol I have adjusted this. Does this correspond to your suggestion?

saucecontrol replied:
You don't need `Unsafe.As` there.  If you were getting errors on a normal cast from `ulong` to `uint`, you might just need an `unchecked` context, although I don't believe that's necessary in CoreLib.

deeprobin replied:
done

deeprobin replied:
Are you sure that this conversion is a good practise :smile: 

The tests are failing because of this...

`System.ArgumentException : Object of type 'System.UInt64' cannot be converted to type 'System.UInt32'.`

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61558-merge-4697f643e9194882ac/System.Runtime.Extensions.Tests/1/console.81399b4e.log?sv=2019-07-07&se=2021-12-06T21%3A26%3A52Z&sr=c&sp=rl&sig=bquVfSnssQk3D6h%2FTnRoDZempuihIEdiWeyeBHkNYEs%3D

tannergooding replied:
The assert is in a test because you are passing in `ulong` to a test that is taking a `uint` parameter:
```csharp
        [Theory]
        [InlineData(0, 0, 0)]
        [InlineData(0, 120, 44758)]
        [InlineData(0, ulong.MaxValue, 42721)]
        [InlineData(123, ulong.MaxValue, 27055)]
        public static void BitOps_Crc32C_ulong(uint crc, uint data, uint expected)
```

My guess is this should be `public static void BitOps_Crc32C_ulong(uint crc, ulong data, uint expected)`

deeprobin replied:
Oh that was a mistake by me 😅 Thank you

deeprobin replied:
@tannergooding Do you know why this does not work on ARM64?

kasperk81 commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +708,166 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+
+
+            // Software fallback
```
doesn't produce the same result as Sse42.Crc32 with `Crc32(1164749927, byte.MaxValue)`

deeprobin committed:
Rename Crc32 to Crc32C

deeprobin committed:
Remove unnessecary heap allocation

deeprobin committed:
Fix software fallback using table-based-CRC

deeprobin committed:
Usage of uint32_t __crc32ch (uint32_t a, uint32_t b)

deeprobin committed:
Remove Crc.Sse42 implementation for Crc32C(uint crc, ulong data)

There is no implementation which returns `uint32_t`. See https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#techs=MMX,SSE,SSE2,SSE3,SSSE3,SSE4_1,SSE4_2,AVX,AVX2,FMA,AVX_VNNI,AVX_512,KNC,AMX,SVML,Other&text=crc

deeprobin committed:
Add SSE 4.2 x64 implementation with byte truncation for Crc32C(uint crc, ulong data)

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -41,6 +41,42 @@ public static class BitOperations
             19, 27, 23, 06, 26, 05, 04, 31
         };
 
+        private static ReadOnlySpan<uint> CrcTable => new uint[]
```
Is this good practise or is using a static heap allocated array a better alternative?

gfoidl replied:
This will allocate the `uint[]` array on each access. The Roslyn optimization (if you had that in mind) only works for one byte sized primitive types at the moment (due to endianess concerns).

> static heap allocated array a better alternative?

At the moment yes, so
```suggestion
        private static readonly uint[] s_crcTable => new uint[]
```

deeprobin replied:
Is it still the rule to use the prefix `_s` in non-legacy code? And would stack allocation still make sense?
Something like: ```cs
private static ReadOnlySpan<uint> CrcTable => s_crcTable;
```

gfoidl replied:
To which stack allocation do you refer? Here's a GC-heap allocation, and that's fine (here) as it's a one-timer (due static readonly).

Allocating the constants on the stack requires a lot of stack spaces, and would be needed to be done for each invocation of the method (or at least for the software fallback) and is not very efficient.

deeprobin replied:
👍🏼

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +744,156 @@ public static nuint RotateRight(nuint value, int offset)
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(ushort)];
+            Unsafe.As<byte, ushort>(ref bytes[0]) = data;
```
```suggestion
            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
```

deeprobin replied:
What benefit does this provide? Does this generate different JIT assembly code?

I prefer in readability the `Unsafe.As` variant and `BitConverter.GetBytes` also uses the `Unsafe.As` approach.

deeprobin replied:
The generated JIT Asm is the same. See [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgAYBYAKDIEYbjyACYugOgCUBXAOwwCW+GOwCS/GFAgAHAMqSAbgLAxcAbkYtxGSb2wAbLn0HD2AYQj5pA/ZPlQlK9TUYpmAMQFRcGABTdcAAtoDGYAE2wMbABKZgBvGmYk5llpbF4AHmAATx0APmYcnVxmAF5mH2wwAGsDfQgwQtyYAG1cAQAvGAgAM38gkOiAXQ1qZOYAVV5cbB6RAEFcLOa0ZgDgqAw831gepuKW8iHY8oio0YBfF1o3eUheMP6N0LOY+MTk1PTl/P3VMoqURqdQafzanW6fXWgxGHySUxmc3YAHUoAIdFMDAIAOa8GCPXbMACyMHw0GyROw3kCBnYAHEYBhODA5rBeCpfEVVNFVq9opcgA=),

gfoidl replied:
In current implementations of JIT it's the same generated code. With `WriteUnaligned` it's explicit that an unaligned write needs to be done, so it's better code-hygiene and definitely future proof iif the JIT would ever emit different machine code for `Unsafe.As`.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +744,156 @@ public static nuint RotateRight(nuint value, int offset)
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(uint)];
+            Unsafe.As<byte, uint>(ref bytes[0]) = data;
```
```suggestion
            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
```

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +744,156 @@ public static nuint RotateRight(nuint value, int offset)
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(ulong)];
+            Unsafe.As<byte, ulong>(ref bytes[0]) = data;
```
```suggestion
            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
```

gfoidl commented on src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:
```diff
@@ -887,5 +887,49 @@ public static void BitOps_RoundUpToPow2_nuint_64(ulong value, ulong expected)
+        [InlineData(123, byte.MaxValue, 1164749927)]
+        public static void BitOps_Crc32C_byte(uint crc, byte data, uint expected)
+        {
+            var obtained = BitOperations.Crc32C(crc, data);
```
Use the concrete type instead of `var`.

Same below.

deeprobin replied:
done

deeprobin committed:
Replace Unsafe.As with unchecked-uint cast

gfoidl reviewed (commented):
Some note -- not for implementation of algorithm though.

deeprobin committed:
Usage of WriteUnaligned instead of As

Co-authored-by: Günther Foidl <gue@korporal.at>

deeprobin committed:
Usage of WriteUnaligned instead of As

Co-authored-by: Günther Foidl <gue@korporal.at>

deeprobin committed:
Usage of WriteUnaligned instead of As

Co-authored-by: Günther Foidl <gue@korporal.at>

deeprobin committed:
Usage of explicit types

deeprobin committed:
Merge branch 'issue-2036' of https://github.com/deeprobin/runtime into issue-2036

deeprobin committed:
Fix Software-Fallback Table to Castalogni equivalent

deeprobin commented on src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:
```diff
@@ -887,5 +887,49 @@ public static void BitOps_RoundUpToPow2_nuint_64(ulong value, ulong expected)
+        [InlineData(0, 120, 44758)]
+        [InlineData(0, ulong.MaxValue, 42721)]
+        [InlineData(123, ulong.MaxValue, 27055)]
+        public static void BitOps_Crc32C_ulong(uint crc, uint data, uint expected)
```
currently failing - see tests

kasperk81 commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +744,155 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+
+
+            // Software fallback
```
still doesn't produce the same result as Sse42.Crc32 with `Crc32(1164749927, byte.MaxValue)`

deeprobin replied:
I know 😦... Do you know why? [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmABgFgAoUgRmqLIAIjaA6AJXwwEsBbGGwCSeDFB54AzjzCS2ADQAcSANzVqeAIYDJAB01gYTACoxJGJgG9qTWyxIBmK3ZeuXNty1pImOCRgBtAF0mMCgwY01gABsjAF4mPBgAd19/YI83aypPVzIEMjQmfIAxEiQAIUUHMgci/IBRWgcKgHYyEtb6hGaAVg6HEvRihABhVoBOAEEJ1toS7odektpaFFpR7vKp2gbWhsVugBEUUanUBoq0TNyRxSmjid7FUYWR1sUKkiOK0c2RyoNEgkaqHEYTCaKaYVFrdFAnByvI6FEYVMqKU5HOojKbjT5lLojXpNByIkjoG65fK0MjE8ZIN6NRy9FAoJD/UreXptSGLMgNBwoCZXEZHVqjFD9dpbXpTEpY1rY/IOBllMiKQn5CUTFBTBwY645W62fLTDUTEpTFH5JCKBqjDZ67qtCq/BolXqahBQsXslndXpHWjI/HdOWtW2ujkIX7lD4oMH5FANeElCooK6Uzz5cgVI52o4NY7fJBkI7l7oO5QlIETRYOBpio4kKbdPYkOYTB3dWgs0bTUYkbpIDZICYq7pDVr7PW9Q3Gk0IKbneEj2gBxQdXq9WhDkYZ5bnBpIboVccoEjEucAo419b87oWp6DsvdF6jXoOeZ194NKaXEhLXnBdlX5QVhXXEZB0UF4pi+Y5g2qHUTxGRwYQqKZJUnacJkLf8A13QMHh7VAWwaXof0aC8Jm+WVgONfIMIcHZaH2OFFB3GjWiOAMKjIbxaChMMJiQf8QVbd4jitZQRzfEpRj43okD3U1+zRMcUJtd0yFaCi5yzNwk1Yipg1+U8vzICZLkZJcyBQMgkBOEV8l6YEKgaFNE29JAmnuJ13gTK1pmtZBuLlKZ/XBblWKeLoDLyMYMPJT0lQQBwuy7JBpVQmZx2eYZ8nzMoKltL1Rn1Zt2Uncilm8GzG1qKYSGFHsSltEhaF5eKXBcwipKxMNBXI3CQr43Zl3JOEO1crjH0qIYjj+Yd4VOe0bOnDEHCxZzvWJNMBUObq7EKtFRkUDro2a2Z+PmRZcVaMpLkre5xl2LzRxHfpUo8ssRLIEL5UDJBlgKhAShVY9Ono24bVHcdlMfZigwcyD8k3eUBSeZ1mqQUsRLDYHRjs7iAxQBwPmnIt9wjadAtPfpXg+OsjsXS55gU4iRk61pWXuKn8iajMKjokYSgxcGOMohASDO7cFJUhBCw828TMrH4gQla9lUBuU2WhqkEE7Dj7J4u5Bi29p+YQCYyD1ckgIBGX7geqWSrIGWvwV1kx3s1YA0VYUGy1pdt0cfkTxZ0X0xpP4FZoqZFHk1GeimVpU/K4ODlWEpwc0tLRh+YUKygy5+MUQtjkLfkvgknMxZK87maNBjvUApAWg+Z1LJ+DUdvbguM0Tx9aHOM6XjhJSzlGLLT3aVo29BzDPUWm2A375Zx317NW7ajvJ0w5MyHTNsJjWZp7J7NoShtsnKzTEGE8Wcs9jTnaSAOA5JRsxbyJ03pNkjqaFsicHRelLOVB6axnTpX1KMBoIVFDzHRgKAMn5yh8S8qnCiBwj6nnsqfNENk2TKUhPJLehkegahHvCLyf4R4dh+JOHUDwLw2S+L2ee0ZbyXFJOMLY+Yj5kAlIsBqZMaiVgUr8WodRAFLnJK6WUed+jkBVBFOE5MhHbmTiZWg6ZgYK2BkGNyUtZgYR2BhN8yYIrkithGRBQZ4TkISiCGijYxzHDlCJY8tcxjTAhI8K2LRBy7Fwm2Eo6o04Rh7Ecdu19dJtlJLbRBk4PwwUhEOWR7DPQyylvZfYKwzgBgcO/d2DYwxrGeM0U2+RdKJ3IjEt8+xiFHGDjqdMTwOrDkcisZSGTm4wzSu+FsDlKzLjTjbYOjxOiyhWFsM8BcLQKxzkpPMnVjhbQbCPHsx5FpKzbETC4jYnE9TGBVcowiRhfm8f/L0blgbt1SsicsLTxH7mMqZScWVyrfFoa5P4u5oysRQDzYECxZFChZPsXBqIYQ5xDGGas/IyFEitAmW8Nk7RwIaKfUG4oIqiVYsOZZEIUHggCSPeBxzjqKw2U0OCWxdHKGAYsd+UJezRiJtUQiNk1hCgjMfEYx5QXci9B6eeDlb7vE6sbQosiKK2zPKgMMJVmjnC9FUL8/ZN77giphGYIUrI23fi6YcrwMJii8g9OCxJx53EWmWIqVLWaJITlsrmvYvgmSlrmfMRzRYOVpO5K25JlyDFpMcdkuiDmVgomsHOCtyaWj+LM2RY5sXhJaY+KoHYtp5xgs2TcwJnRwRtsU5O/5FRfJ8f/QSKZ+xwiGIgxU1SYzNnSmTIssjc6Q3iVzWk9oolc0vDuMWCt4ENi7Mky5KAhhi1EpWMcnQSpSxaa8NJLaOyYSPlZJ17xhSKhdD4z4i10xnkfDBCUPMW2iT2BKZMp4gRSVtsHU4bJHDnADD05ErkwxNk/DueKABfNQ/T3CgbsLocQAA3TQGAjCsB8H4UQTAADKEAABmGBkiaFgCUTQ0RojAAMAAawABRIYsGEMARRgAAE84NMAACawc0AASisJHfwTAMBRFiCIRjMAEBMASKR/wrHSOkao0wAAekxlj7GABkosSisZAwuUI4RhMaYiLxmAAQeMxBgPxwTIRZOSc0wAPgs0wRQqn1DgbcEQVo2mZPKZzu5koanPCAfs7kUgJAOMOdcBRlzCRVioB5hCDsXnjR0YY8xnjWm4uCAALKaAQAANXwzgGAMXbiRxg1AJgkhJAwCy9EHLWmUOlYvGwUY4Rinmeo3JnjdmgsuEK8VjDWGcMwDwwRojYBiNVe69h3D+HCMkaa0UBLbG8t+c6qRgAJAAIhQyhhoTAUBsBICAKwJWyvZZgIBlbbX1OsAmMttbo3etMH65Nobe3LCSBu+NgbJGTtne8zcHzVBANAA=)

kasperk81 replied:
dropping `^ 0xFFFFFFFF` mask fixes it. with backing table, you have computed value, no need for extra work.

deeprobin committed:
Fix test signature

deeprobin committed:
Remove mask for software fallback

kasperk81 committed:
reuse reflected table generator

deeprobin committed:
Fix test data

kasperk81 commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -41,6 +41,42 @@ public static class BitOperations
             19, 27, 23, 06, 26, 05, 04, 31
         };
 
+        private static readonly uint[] s_crcTable = new uint[]
```
you can reuse the existing table generator with non-revered polynomial value `0x82F63B78u`
https://github.com/kasperk81/runtime/commit/d74d96a9002e6691ef8d501c80e6cd7669db5097

deeprobin replied:
@kasperk81 Build failures `Type 'Crc32ReflectedTable' is a static holder type but is neither static nor NotInheritable`

kasperk81 replied:
add static to the class? you don't had to revert the whole thing.. take your time and build and test it locally

kasperk81 commented on src/libraries/System.Runtime.Extensions/System.Runtime.Extensions.sln:
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
```
revert changes in both .sln files, as slngen pr will overwrite them

deeprobin committed:
Usage of CRC Table Generator instead of constant values

deeprobin committed:
Fix solution file

kasperk81 commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -41,41 +41,33 @@ public static class BitOperations
         };
 
-        private static readonly uint[] s_crcTable = new uint[]
+        private static readonly uint[] s_crcTable = GenerateReflectedTable(0x82F63B78u);
```
i meant reuse the code instead of duplicating it https://github.com/kasperk81/runtime/commit/d74d96a9002e6691ef8d501c80e6cd7669db5097

kasperk81 commented on src/libraries/System.Runtime/System.Runtime.sln:
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
```
also revert changes in this file

deeprobin committed:
Revert "Usage of CRC Table Generator instead of constant values"

This reverts commit d44a2151b917a1650d844da98f82d50bfad479ce.

deeprobin committed:
Merge pull request #1 from kasperk81/issue-2036

reuse reflected table generator

deeprobin committed:
Fix solution file

deeprobin committed:
Merge branch 'issue-2036' of https://github.com/deeprobin/runtime into issue-2036

kasperk81 commented on src/libraries/Common/src/System/Numerics/Crc32ReflectedTable.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace System.Numerics
+{
+    internal class Crc32ReflectedTable
```
```suggestion
    internal static class Crc32ReflectedTable
```

deeprobin committed:
Make Crc32ReflectedTable static

Co-authored-by: kasperk81 <83082615+kasperk81@users.noreply.github.com>

AndyAyersMS commented:
@dotnet/jit-contrib -- is there anyone with arm64 hardware that can help drill into these failures?

```
    System.Numerics.Tests.BitOperationsTests.BitOps_Crc32C_ulong(crc: 0, data: 120, expected: 3511526341) [FAIL]
      Assert.Equal() Failure
      Expected: 3511526341
      Actual:   2142603813
      Stack Trace:
        /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(932,0): at System.Numerics.Tests.BitOperationsTests.BitOps_Crc32C_ulong(UInt32 crc, UInt64 data, UInt32 expected)
      Assert.Equal() Failure
      Expected: 3293575501
      Actual:   1147535477
    System.Numerics.Tests.BitOperationsTests.BitOps_Crc32C_ulong(crc: 0, data: 18446744073709551615, expected: 3293575501) [FAIL]
      Stack Trace:
        /_/src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs(932,0): at System.Numerics.Tests.BitOperationsTests.BitOps_Crc32C_ulong(UInt32 crc, UInt64 data, UInt32 expected)
      Assert.Equal() Failure
      Expected: 3460750817
      Actual:   3042741683
      Stack Trace:
    System.Numerics.Tests.BitOperationsTests.BitOps_Crc32C_ulong(crc: 123, data: 18446744073709551615, expected: 3460750817) [FAIL]
```

echesakovMSFT commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +710,155 @@ public static nuint RotateRight(nuint value, int offset)
+            if (Crc32.Arm64.IsSupported)
+            {
+                // uint32_t __crc32d (uint32_t a, uint64_t b)
+                return Crc32.Arm64.ComputeCrc32(crc, data);
```
Shouldn't this use `Crc32.Arm64.ComputeCrc32C`?

deeprobin replied:
Yes. It should. Thank you, fixed in last commit.

deeprobin committed:
Fix wrong intrinsic for Arm64/BitOperations.Crc32C(uint crc, ulong data)

deeprobin committed:
Merge branch 'issue-2036' of https://github.com/deeprobin/runtime into issue-2036

deeprobin commented on src/libraries/Common/src/System/Numerics/Crc32ReflectedTable.cs:
```diff
@@ -0,0 +1,34 @@
+            {
+                uint val = (uint)i;
+
+                for (int j = 0; j < 8; j++)
```
I think this part could be SIMD-vectorized, if supported on CPU. There are CPUs that support SSE and SSE2 but not SSE4.2. For this case this is relevant in my eyes.

## Example intrinsics

### SSE
```h
__m128 _mm_and_ps (__m128 a, __m128 b);
```

### SSE2
```h
__m128i _mm_and_si128 (__m128i a, __m128i b);
```

### AVX
```h
__m256 _mm256_and_ps (__m256 a, __m256 b);
```

### AVX2
```h
__m256i _mm256_and_si256 (__m256i a, __m256i b);
```

saucecontrol replied:
It's a one-time cost, and a small one at that.  Not worth the added complexity even if it could be vectorized.

deeprobin replied:
That it is a one-time cost is true. But vectorizing it would not be a big effort in my opinion, since we have Vector* classes.
Would a vectorization cause any other negative effects code-wise?

AndyAyersMS replied:
Depending on what you do, it might block the method from being prejitted.

deeprobin replied:
Oh okay then I will leave the vectorization for now. If at some point this method should really lead to a bottleneck, it can still be changed.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +710,155 @@ public static nuint RotateRight(nuint value, int offset)
+
+            // Software fallback
+
+            Span<byte> bytes = stackalloc byte[sizeof(uint)];
+            Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(bytes), data);
+
+            foreach (byte b in bytes)
+            {
+                int tableIndex = (int)((crc ^ b) & 0xFF);
+                crc = s_crcTable[tableIndex] ^ (crc >> 8);
+            }
```
These `stackalloc`s are inefficient and unnecessary.  Why not just operate directly on the value passed in?

And a known-size loop this small is probably best unrolled manually.  Something like [one of these](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCEBXAMy5il3YBLAHbYoATwDcNek1YAlDsIyD8MFgEllfCAAcAynwBugsDFzTajZi0XLV6gMIR8uwQBs+hqCbMWaMgDMTKQMjgwA3jQMMQwcIhgA2gC6DLgA+mBQYAAq2MCeDAC8DMIwAO5xCYmkAKxIyZaxDNGxxMHxymHZgaSOABSdGAxZYGhVXQAm2BjYAJStMVHUzc36utjCADzA4hgwAHwMu/u4xWmzYADW2O7uEGDHezCJuIIAXjAQXIMJc42LVYAVWEuGwvBYAHUoIJ9iDboIAOZlSb9WBcBgAWRg+Gg4kxYlwAAtbiwAOIwDDyGC8WDCMz9E7mObjaazOaWQHNLjQGDYMBEhiM57HBgiJ6nBYrVZLLkysVdWYFGBaSYwBDnfp/fr9UYMAB6xzmDAAZAw6AgAGKWjlymV6koZUZ5ZWJJWeVXq1KG3XZBgHI4ADlt0plAF8AqHVsQAOwjbJNWIR6iA9oTYaOHp9Ui/LqjcZDBhs+aA5bysUY/psWHOYRGPj7KCaXAAGVhGE8AFFhJNBJspeXmsXztXRBIAAow/CwwT1gTU+v8GDd3ubMq4XD9Yshu0MdHphjuDjDEr77G4iQE/gk9zkykAQSgUGw4gAIjNsNTaTB6TB+k7shdTwQ3LXceSgIUEjFc46EkaCtjSD4vh+IYOTFABqdCB3lMtBxiB0GBBMEIXvSZUX3I8MHGfphFQ4V9jmX1Hh9Jk5m3Y0fT1AMGGDRM8OHAMSkDPjVmTXdY3jMARLEqM00LTMwF6RxAlzYZ8wPbdS13QRK2rDBa0XRtmzbDAO2XHs+2EbD+I/EcRDEcRJ1UGc51sGBF1wczV2EddN23Tko2afdC0o84zxxPEr2JUkKQwR9nzfD8vz4H8GQA3J8mAkTml3AiiPBdRSPImlD2PajaO1VimINIVWPYmrqu43jdwEg4hOy2I8tBAqWCKtESso8q6KqvUWOeNiPzmDihS4oMQPlVr2tyv0SnykiyP6jFBqFCrlEYka/TGhj6s4v0mvmmVFp4jr8JWwjuvW4qtrKnbhvG6rDROmazrmgLywk0ZpJoMMgA==) (untested -- endianness might be backwards, but you should get the idea)

deeprobin replied:
> These stackallocs are inefficient and unnecessary. Why not just operate directly on the value passed in?
Can you explain me, why this is inefficient?

> And a known-size loop this small is probably best unrolled manually.
Why doesn't the JIT unroll loops this small automatically?

saucecontrol replied:
> Can you explain me, why this is inefficient?

At present, the JIT always emits a stack buffer overrun check when you `stackalloc`.  Since your goal is only to write a single primitive value to the stack, there are multiple ways to do that without using `stackalloc`.  For example, you can create a span directly over the data argument like this:

```cs
ReadOnlySpan<uint> span = MemoryMarshal.CreateReadOnlySpan(ref data, 1);
```

Note the [codegen difference](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8YUCAAcAymIBuvMDFwBuGvSatOPfoIDCEfBN4AbMTKjzFKmmoDMTUg10MA3jQYeGHXjwDaAXQZcAH0wKDAAFWxgUwYAXgYuGAB3Lx8MX1IAViR/VWpPQoZ3T2J7bx4ncNtSXQAKCowGMLA0NMqAE2wMbABKEo83AqKPKQlsLgAeYABPDBgAPgZZ+dx4oJ6wAGtsY2MIMGW5mF9cXgAvGAgAMwb03ryBooBVLlxsa8EAdSheedfdrwAOZJDp1WDXBgAWRg+GgMyh2CguAAFrsWABxGAYNgwT6wLiKOorJS9NpdHq9fJPQrXaAwbBgFEMYnHZYMHxHVb9YYjIYjEbpBg9GIwEQdGAIdZ1e51OotBgAPWWvQYADIGHQEAAxbVUmkChUJEItKKi3wi0ziyWBZXy8IMBZLAAc+t5RQAvjZ3YViAB2ZrhfKFL3UJ5ldpNXTVWrBAByECkmy2AEF9gc7pUWm1GgwKX0nvyBbjsB0APJcYwzMYTSaNJa4cZcdYwuFQBFI1Ho6MM+Yl8uV6tN8F4vPdbBtchug0eOmwRnM1nzdmc1vwxHItHGFgp3AAIWOuDqjYmvR5AtcM6KQstYq4EqlCRlPF6coVyuAqo1Wt1bovhSNIJQnCM1TAtaIrXvG0lRZBUnQYV1gwFUMr39QMwCQjxQw9IA) with just that change.

> Why doesn't the JIT unroll loops this small automatically?

JIT has to balance the time it takes to identify and realize optimizations with the benefit they offer.  Loop unrolling is not always a win, and the JIT doesn't have the same amount of time, say, an optimizing C++ compiler does to evaluate that.  There is one place JIT will auto-unroll a loop because it's likely to always be of benefit, and that is loops over `Vector<T>.Count`, `Vector64<T>.Count`, `Vector128<T>.Count`, and `Vector256<T>.Count`.  You could take advantage of this fact to force an unroll of 4 with e.g. `for int i = 0; i < Vector128<int>.Count; i++`, but then in your code, you'd introduce bounds checks using `i` to index into the span.

The samples I put up yesterday eliminate both of these problems and remove the bounds checks for the CRC lookup table, so they would be significantly faster.

deeprobin replied:
Thank you for the clarification. Do you know why the JIT method ends with an `int3` instruction in both cases? Is it mandatory?

saucecontrol replied:
The `call` that precedes the `int3` in both of those methods is the exception for the bounds check.  I believe JIT inserts the `int3` to indicate that the method does not actually return in that case.  Of course if you eliminate the bounds checks into the LUT, you also eliminate the code to throw on out of bounds, which is of course not possible when looking up in a 256-element array with a 1 byte index.

deeprobin committed:
Reduce amount of stackalloc statements and replace them with MemoryMarshal.CreateReadOnlySpan

deeprobin committed:
Loop unwinding and performance optimization

deeprobin committed:
Merge branch 'dotnet:main' into issue-2036

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +711,168 @@ public static nuint RotateRight(nuint value, int offset)
+
+
+            // Software fallback
+            int tableIndex = (int)((crc ^ data) & 0xFF);
```
It's worth using `GetArrayDataReference` and `Unsafe.Add`ing the index here as is done in the other methods.

BTW, I'm assuming you got this reference implementation from a C sample?  `& 0xFF` is necessary in C to isolate the low byte, because C doesn't guarantee the upper bits will be cleared.  Most C compilers will optimize this to a zero extension (e.g. `movzx`) where possible.  JIT will take your code literally and emit the `and`.  A cast to `byte` (as used in the other methods) is sufficient to clear the upper bits because .NET guarantees this where C doesn't.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +711,168 @@ public static nuint RotateRight(nuint value, int offset)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+
```
Nit: remove extra newlines

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +711,168 @@ public static nuint RotateRight(nuint value, int offset)
+
+            crc = Unsafe.Add(ref lut, (nint)(byte)(crc ^ (byte)data)) ^ (crc >> 8);
+            data >>= 8;
+            crc = Unsafe.Add(ref lut, (nint)(byte)(crc ^ (byte)data)) ^ (crc >> 8);
```
```suggestion
            crc = Unsafe.Add(ref lut, (nint)(byte)(crc ^ data)) ^ (crc >> 8);
```

Cast here is unnecessary because the the upper byte was cleared by the shift above.

deeprobin committed:
Remove unnessecary cast

Co-authored-by: Clinton Ingram <clinton.ingram@outlook.com>

deeprobin committed:
Use MemoryMarshal.GetArrayDataReference instead of direct array access

deeprobin committed:
Remove unnessecary newlines

deeprobin committed:
Merge branch 'issue-2036' of https://github.com/deeprobin/runtime into issue-2036

deeprobin committed:
Merge branch 'dotnet:main' into issue-2036

deeprobin commented:
Most of the CI checks are sucessful. The remaining are some external issues.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -41,6 +42,8 @@ public static class BitOperations
             19, 27, 23, 06, 26, 05, 04, 31
         };
 
+        private static readonly uint[] s_crcTable = Crc32ReflectedTable.Generate(0x82F63B78u);
```
You might want to change this initializer to call a method that bails out and returns `Array.Empty<uint>()` if the hardware instruction is supported, so you don't take the cost of building the LUT when it won't be used.

kasperk81 replied:
```suggestion
        private static readonly uint[] s_crcTable = Sse42.IsSupported || Crc32.IsSupported ?
            Array.Empty<uint>() : Crc32ReflectedTable.Generate(0x82F63B78u);
```
will be handled by R2R the same way as other hardware intrinsic code with software fallback.

deeprobin replied:
@kasperk81 This results in test failures ... (see CI) :/

saucecontrol replied:
That's because although `Sse42.IsSupported` is true in the x86 CI environment, `Sse42.X64.IsSupported` is false.  The `ulong` overload fails that check and goes for the fallback code, only to find the LUT empty (this could have easily resulted in AV rather than just bad CRC values).

What you want to do in those cases is decompose the `ulong` down to two `uint` values and call the 32-bit intrinsic instead of falling back to the LUT version of the code.

deeprobin committed:
Update src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

Co-authored-by: kasperk81 <83082615+kasperk81@users.noreply.github.com>

kasperk81 commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -41,6 +42,9 @@ public static class BitOperations
             19, 27, 23, 06, 26, 05, 04, 31
         };
 
+        private static readonly uint[] s_crcTable = Sse42.IsSupported || Crc32.IsSupported ?
```
```suggestion
        private static readonly uint[] s_crcTable = Sse42,X64.IsSupported || Crc32.Arm64.IsSupported ?
```

deeprobin replied:
Thank you

saucecontrol replied:
This is not a good fix, because it means the LUT will always be created on x86, even if it is never used.  The `ulong` overload will also always be processed byte-by-byte with the fallback implementation, when it would be *much* faster to decompose to 2 `uints` and call the intrinsic.

deeprobin replied:
Do you have an idea to provide a better fix? Can't something like this be optimized by the R2R "compiler"?

saucecontrol replied:
The potential R2R complication I was referring to in a previous comment was related to this: https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/vectors-and-intrinsics.md#code-review-rules-for-code-written-in-systemprivatecorelibdll

The rules are different under `crossgen2`, so I don't believe that's a problem anymore.

You've got 2 problems left:
1) x86 clients will get the slow software fallback if they call the `ulong` overload, when it would be faster to break it down and call the `uint` intrinsic twice.  You can fix this by adding a 32-bit `Sse42.Crc32` fallback path when `Sse42.X64` is not supported but `Sse42` is.
2) If neither the ARM nor SSE4.2 intrinsic is supported, the CRC LUT will be created as part of the `BitOperations` class initializer.  Since other `BitOperations` methods are called from various places in the BCL, this means the LUT will always be created regardless of whether `BitOperations.Crc32C` is ever used.  The two ways to fix that would be:
a) Lazy-initialize the LUT only when it's needed, or
b) Move the static LUT field to its own static class, so that the initializer for that class only runs if the fallback LUT is actually needed.
The lazy-initialization option would incur a cost on each call to `Crc32C`, so it would likely be more desirable to take the latter option.

kasperk81 replied:
> If neither the ARM nor SSE4.2 intrinsic is supported,

this scenario (someone on ancient computer getting slightly degraded performance) is not something we should worry about when rest of the runtime doesn't either.

saucecontrol replied:
It's not just about ancient hardware.  WASM doesn't support either ISA, so it would be included.

kasperk81 replied:
one kb lut field, which we are talking about here, is drop in a bucket compared to what a simplest wasm app's startup loads. 
but why not, if you think calling intrinsic on decontructed input values and reconstructing the output for supported endianess will produce the same result more efficiently, feel free to suggest the code change like i did above.

saucecontrol replied:
To use your metaphor, with enough drops, the bucket will overflow.  I won't speak for the BCL team, but it's clear they have made a concerted effort to reduce startup overhead in the last couple of .NET releases.  If they want to take this PR as is, that's fine -- I'm just pointing out an easily avoidable overhead this would introduce.

> feel free to suggest the code change like i did above

It wouldn't be a one-line change, and this is not my PR.  Leaving it as a learning exercise for @deeprobin

deeprobin replied:
Yes I still have a lot to learn :D. In this regard, a quote from Winston Churchill.

> I am always ready to learn although I do not always like being taught.

deeprobin committed:
Add x64 SSE check

Co-authored-by: kasperk81 <83082615+kasperk81@users.noreply.github.com>

deeprobin commented:
Can somebody re-run the failing CI-workflows 😄? thanks

deeprobin committed:
Move Crc32 Software into own class

deeprobin committed:
Fix merge conflict

deeprobin committed:
Remove IsSupported check

deeprobin committed:
Fix parameter naming

deeprobin committed:
Fix fallback implementation method naming

deeprobin committed:
Improve documentation

deeprobin committed:
Move bswap into Crc32Fallback class

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +709,195 @@ public static nuint RotateRight(nuint value, int offset)
+            {
+                if (!BitConverter.IsLittleEndian)
+                {
+                    data = BinaryPrimitives.ReverseEndianness(data);
```
No harm in it, but this isn't necessary for the `byte` overload, as you can't swap the order of a single byte.  As the docs say:
> This method effectively does nothing and was added only for consistency.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +709,195 @@ public static nuint RotateRight(nuint value, int offset)
 #endif
         }
+
+        private static class Crc32Fallback
```
Nit: I'm not sure about the BCL style preferences around nested classes, but I'd personally put this at the end of the parent rather than having it hanging out in the middle. Area owners might prefer this to be split into another file, though 🤷‍♂️

deeprobin replied:
@tannergooding What do you think 😄?

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +709,195 @@ public static nuint RotateRight(nuint value, int offset)
+
+        private static class Crc32Fallback
+        {
+
```
Nit: remove extra newline.

saucecontrol commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +709,195 @@ public static nuint RotateRight(nuint value, int offset)
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, ulong data)
+        {
+            if (Sse42.X64.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                ulong result = Sse42.X64.Crc32(crc, data);
+
+                return unchecked((uint)result);
+            }
```
```suggestion
            if (Sse42.X64.IsSupported)
            {
                return (uint)Sse42.X64.Crc32(crc, data);
            }
            if (Sse42.IsSupported)
            {
                return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), (uint)(data >> 32));
            }
```

This is what I meant about breaking down the `ulong` overload for x86.  Use the 64-bit instruction when possible, but rather than falling back to the software implementation, just break it into halves and call the `uint` intrinsic if it's available.

Same can be done for ARM, except SSE4.2 implies little endian whereas ARM is technically bi-endian so you must byte swap BE input if you want to process in LE order to match the above.

deeprobin committed:
Implement efficient usage of SSE x86/x64

Co-authored-by: Clinton Ingram <clinton.ingram@outlook.com>

deeprobin committed:
Style Changes for BitOperations.cs

deeprobin committed:
Remove unnessecary `bswap` of a single byte

SingleAccretion commented:
FYI I have filed #62692 on the Jit assert (and more).

I believe this change can in the meantime be unlocked by using a temporary:
```cs
var a = (uint)(data >> 32);
return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), a);
```

deeprobin commented:
> FYI I have filed #62692 on the Jit assert (and more).
> 
> I believe this change can in the meantime be unlocked by using a temporary:
> 
> ```cs
> var a = (uint)(data >> 32);
> return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), a);
> ```

Thank you very much.  I would wait for Tanner's review first and then we'll take a closer look at how to proceed.

/cc @tannergooding

tannergooding commented:
I will take a more in depth look at the JIT asserts next week when I have time.

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -708,5 +709,190 @@ public static nuint RotateRight(nuint value, int offset)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static uint Crc32CCore(ref uint lookupTable, uint crc, uint data)
+            {
+                crc = Unsafe.Add(ref lookupTable, (nint)(byte)(crc ^ (byte)data)) ^ (crc >> 8);
```
Should we add a comment here, why a loop is not used here?

tannergooding commented:
@deeprobin, I'd recommend marking this as "draft" for the time being.

I've got a number of other things that are taking precedence over looking at the Crc32 issues in the JIT at the moment.

deeprobin commented:
@tannergooding This PR was auto-closed. Can you re-open it :smile:?
Work is done but blocked by #62692.

danmoseley commented:
@deeprobin I thought PR author could reopen PRs?

Nevertheless I suggest to leave it closed until it is unblocked.

deeprobin commented:
> @deeprobin I thought PR author could reopen PRs?

Nope, I have only the "Delete branch" button

> Nevertheless I suggest to leave it closed until it is unblocked.

Okay 👍🏼

EgorBo commented:
Re-open per request

deeprobin commented:
merge incoming

deeprobin committed:
Merge branch 'main' into issue-2036

deeprobin committed:
Merge branch 'main' into issue-2036

deeprobin committed:
Fix doc comments

deeprobin commented:
There was probably an failure during the merge -> fixed it

Now this PR should be ready. Can we still get this in .NET 7, since before this PR was already heavily reviewed and it is imo risk-free to merge this.

/cc @tannergooding @danmoseley

deeprobin commented:
We can remove the `blocked` label btw.

deeprobin commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -51,7 +53,7 @@ public static class BitOperations
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         [CLSCompliant(false)]
-        public static bool IsPow2(uint value) => (value & (value - 1)) == 0 && value != 0 ;
+        public static bool IsPow2(uint value) => (value & (value - 1)) == 0 && value != 0;
```
I'm not sure right now why this change was made (probably automatically by VS), but it doesn't hurt I think.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,191 @@ public static nuint RotateRight(nuint value, int offset)
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * uint32_t __crc32b (uint32_t a, uint8_t b)
+        /// * uint32_t __crc32cb (uint32_t a, uint8_t b)
```
This info shouldn't be in the "summary". Everything after the first sentence should be deleted.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,191 @@ public static nuint RotateRight(nuint value, int offset)
+        /// Accumulates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
```
Same here and all the others like it.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,191 @@ public static nuint RotateRight(nuint value, int offset)
+        /// * uint32_t __crc32cb (uint32_t a, uint8_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
```
"the checksum data" is ambiguous. did you mean "the data for which to compute the checksum"?

stephentoub reviewed (commented):
> it is imo risk-free

There's no such thing ;-)

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,191 @@ public static nuint RotateRight(nuint value, int offset)
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32C(uint crc, byte data)
```
Is there anywhere throughout dotnet/runtime these should be used?

deeprobin replied:
Unfortunately, I am not sure. I first thought of Zip, but it uses the normal "CRC32" and not the Castagnoli algorithm.

Joe4evr replied:
There's #44194 putting `CLSCompliant` up for a rethink, but it looks like there hasn't been any hard decisions yet on when it should/shouldn't be used.

deeprobin replied:
Thanks for the info, however as long as this is not approved I think we should continue with the current scheme.

deeprobin committed:
fix doc

deeprobin commented:
> There's no such thing ;-)

*A small risk ;-)

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -53,7 +53,7 @@ public static class BitOperations
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         [CLSCompliant(false)]
-        public static bool IsPow2(uint value) => (value & (value - 1)) == 0 && value != 0;
+        public static bool IsPow2(uint value) => (value & (value - 1)) == 0 && value != 0 ;
```
please revert this

deeprobin replied:
done

deeprobin committed:
Fix

tannergooding commented:
@deeprobin, can you please update the PR by merging `main` from `dotnet/runtime` into your branch and pushing it up?

That will allow CI to run against the updated head. I'm working on reviewing the code today.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The data for which to compute the checksum</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
```
```suggestion
        [CLSCompliant(false)]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                // uint32_t __crc32b (uint32_t a, uint8_t b)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
```
```suggestion

            if (Crc32.IsSupported)
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        {
+            if (Sse42.IsSupported)
+            {
+                // uint32_t __crc32b (uint32_t a, uint8_t b)
```
These comments don't add much and we don't use them anywhere else:
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32cb (uint32_t a, uint8_t b)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u16 (unsigned int crc, unsigned short v)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint16_t b)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Crc32.IsSupported)
```
```suggestion

            if (Crc32.IsSupported)
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+            if (Crc32.IsSupported)
+            {
+                // uint32_t __crc32ch (uint32_t a, uint32_t b)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                return Crc32.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            {
+                return (uint)Sse42.X64.Crc32(crc, data);
+            }
+            if (Sse42.IsSupported)
```
```suggestion

            if (Sse42.IsSupported)
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            {
+                return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), (uint)(data >> 32));
+            }
+            if (Crc32.Arm64.IsSupported)
```
```suggestion

            if (Crc32.Arm64.IsSupported)
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+                return Crc32.Arm64.ComputeCrc32C(crc, data);
+            }
+
+            // Software fallback
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+            if (Crc32.Arm64.IsSupported)
+            {
+                // uint32_t __crc32cd (uint32_t a, uint64_t b)
```
```suggestion
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The data for which to compute the checksum</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
```
```suggestion
        [CLSCompliant(false)]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+        {
+            if (Sse42.X64.IsSupported)
+            {
+                return (uint)Sse42.X64.Crc32(crc, data);
```
A comment explaining that the instruction returns a 64-bit register with the upper 32-bits set to `0` might be beneficial

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+            }
+            if (Sse42.IsSupported)
+            {
+                return Sse42.Crc32(Sse42.Crc32(crc, (uint)data), (uint)(data >> 32));
```
More readable when broken apart:
```suggestion
                uint result = Sse42.Crc32(crc, (uint)(data));
                return Sse42.Crc32(result, (uint)(data >> 32));
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+
+        private static class Crc32Fallback
+        {
+            private static readonly uint[] s_crcTable = Crc32ReflectedTable.Generate(0x82F63B78u);
```
nit: Would be beneficial to comment that this is `0x1EDC6F41`

tannergooding commented on src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Table.cs:
```diff
@@ -9,32 +11,6 @@ public sealed partial class Crc32 : NonCryptographicHashAlgorithm
-
-            return table;
-        }
+        private static readonly uint[] s_crcLookup = Crc32ReflectedTable.Generate(0xEDB88320u);
```
nit: Would be beneficial to comment that this is: `0x04C11DB7`

deeprobin replied:
there is ![grafik](https://user-images.githubusercontent.com/10373452/192624471-922d9583-4c80-4a46-9e49-3485b57879af.png)

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,175 @@ public static nuint RotateRight(nuint value, int offset)
+
+            internal static uint Crc32C(uint crc, ushort data)
+            {
+                if (!BitConverter.IsLittleEndian)
+                {
+                    data = BinaryPrimitives.ReverseEndianness(data);
+                }
```
Is this correct?

If I define `ushort x = 0x0201`, why should big endian process these as `0x02` then `0x01`? We have an already in register 16-bit value, so any endianness fixups should have been handled by the user already, no?

saucecontrol replied:
Presumably, the caller would be reading the value from memory (in host order), not setting it as a literal.  But it's a good question, since I assume arm64 running in BE mode would read from memory in BE order, but then the crc32 intrinsic assumes LE within the register.  I suppose if the byte order fixup is handled here, it should be done in the outer method and not just in the fallback.

The tests added with this are using literals as well, so I'd assume those would fail on the s390 pipelines as is.  Perhaps a more realistic span-based test would be in order, whether the endianness issue is handled by these methods or by the caller.

tannergooding replied:
> Presumably, the caller would be reading the value from memory (in host order), not setting it as a literal. But it's a good question, since I assume arm64 running in BE mode would read from memory in BE order, but then the crc32 intrinsic assumes LE within the register. I suppose if the byte order fixup is handled here, it should be done in the outer method and not just in the fallback.

The general point is we're taking an input "as is", so it should be presumed to already be in the correct order with any fixups done by the caller. As it is I believe BigEndian will return a different result for the same input as compared to LittleEndian

deeprobin committed:
Merge remote-tracking branch 'upstream/main' into issue-2036

deeprobin committed:
Merge remote-tracking branch 'upstream/main' into issue-2036

deeprobin committed:
Apply suggestions

dakersnar commented on src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs:
```diff
@@ -730,6 +732,164 @@ public static nuint RotateRight(nuint value, int offset)
+        {
+            if (Sse42.IsSupported)
+            {
+                return Sse42.Crc32(crc, data);
```
Is this call using the correct "Castagnoli algorithm" as you referred to earlier?

deeprobin replied:
Yes. See https://en.wikipedia.org/wiki/SSE4 -> SSE4.2

dakersnar reviewed (approved):
Couple clarifying questions, but overall LGTM

deeprobin commented on src/libraries/System.Runtime.Extensions/tests/System/Numerics/BitOperationsTests.cs:
```diff
@@ -887,5 +887,49 @@ public static void BitOps_RoundUpToPow2_nuint_64(ulong value, ulong expected)
         }
+
+        [Theory]
+        [InlineData(0, 0, 0)]
+        [InlineData(0, 120, 4215344322)]
+        [InlineData(0, byte.MaxValue, 2910671697)]
+        [InlineData(123, byte.MaxValue, 1164749927)]
```
It was the trust in this algorithm. More simply, I don't think my CPU is lying to me.

danmoseley commented:
Thank you for the contribution @deeprobin !

dakersnar commented:
One of these tests seems to be failing on tvOS:
https://github.com/dotnet/runtime/issues/76830
https://github.com/dotnet/runtime/pull/76952

deeprobin commented:
I don't think the test is the issue, I think the issue is the Crc32C implementation in clr. These were implemented previously (I think it was Tanner).

Tanner can you locate where and what the issue is?

/cc @tannergooding
